persistence/remote/src/main/java/org/infinispan/persistence/remote/upgrade/MigrationTask.java

```diff
@@ -37,46 +51,97 @@
-      if (storeIterator.hasNext()) {
-         RemoteStore store = storeIterator.next();
-         final RemoteCache<Object, Object> storeCache = store.getRemoteCache();
-         RemoteStoreConfiguration storeConfig = store.getConfiguration();
-         if (!storeConfig.hotRodWrapping()) {
-            throw log.remoteStoreNoHotRodWrapping(cache.getName());
+   }
+
+   private void migrateEntriesWithMetadata(RemoteCache<Object, Object> sourceCache, AtomicInteger counter) {
+      try (CloseableIterator<Map.Entry<Object, MetadataValue<Object>>> iterator = sourceCache.retrieveEntriesWithMetadata(segments, readBatch)) {
+         while (iterator.hasNext()) {
+            Map.Entry<Object, MetadataValue<Object>> entry = iterator.next();
+            if (!Arrays.equals((byte[]) entry.getKey(), ignoredKey)) {
+               MetadataValue<Object> metadataValue = entry.getValue();
+               int lifespan = metadataValue.getLifespan();
+               int maxIdle = metadataValue.getMaxIdle();
+               long version = metadataValue.getVersion();
+               long created = metadataValue.getCreated();
+               long lastUsed = metadataValue.getLastUsed();
+               Metadata metadata = new EmbeddedMetadata.Builder()
+                     .version(new NumericVersion(version))
+                     .lifespan(lifespan, TimeUnit.SECONDS)
+                     .maxIdle(maxIdle, TimeUnit.SECONDS)
+                     .build();
+               InternalMetadataImpl internalMetadata = new InternalMetadataImpl(metadata, created, lastUsed);
+               executorService.submit(() -> {
+                  Object key = entry.getKey();
+                  if (!deletedKeys.contains(new WrappedByteArray((byte[]) key))) {
+                     cache.getAdvancedCache().withFlags(Flag.SKIP_CACHE_LOAD, Flag.ROLLING_UPGRADE).putIfAbsent(entry.getKey(), entry.getValue().getValue(), internalMetadata);
+                  }
+                  int currentCount = counter.incrementAndGet();
+                  if (log.isDebugEnabled() && currentCount % 100 == 0)
+                     log.debugf(">>    Migrated %s entries\n", currentCount);
+               });
+            }
          }
-         final AtomicInteger count = new AtomicInteger(0);
-         ExecutorService es = Executors.newFixedThreadPool(threads);
-         migrateEntriesWithMetadata(storeCache, cache, es, knownKeys, count, segments, readBatch);
-         gracefulShutdown(es);
-         return count.get();
       }
-      return null;
    }
 
    @Override
    public void setEnvironment(Cache<Object, Object> cache, Set<Object> inputKeys) {
+      ComponentRegistry cr = cache.getAdvancedCache().getComponentRegistry();
+      PersistenceManager loaderManager = cr.getComponent(PersistenceManager.class);
+      this.stores = loaderManager.getStores(RemoteStore.class);
+      Marshaller marshaller = new MigrationMarshaller();
       this.cache = cache;
+      this.cache.addListener(listener);
```

Comment:
Ah I see now. In that case you should call `addFilteredListener` instead of normal `addListener`. This way it should reduce a lot of the cluster communications.